pkg/reconciler/pipelinerun/pipelinerun_expired_test.go

```diff
@@ -0,0 +1,132 @@
+		tb.PipelineRunLabel("tekton.dev/pipeline", "test-pipeline"),
+		tb.PipelineRunLabel(pipeline.GroupName+pipeline.PipelineTaskLabelKey, "hello-world-1"),
+		tb.PipelineRunLabel("tekton.dev/pipelineRun", "test-pipeline-run-with-annotations"),
+		tb.PipelineRunAnnotation("PipelineRunAnnotation", "PipelineRunValue"),
+		tb.PipelineRunSpec(
+			"hello-world",
+			tb.PipelineRunServiceAccountName("test-sa"),
+		),
+	)
+
+	if !completionTime.Inner.IsZero() {
+		c := apis.Condition{Type: apis.ConditionSucceeded, Status: v1.ConditionTrue, LastTransitionTime: completionTime}
+		pr.Status.Conditions = append(pr.Status.Conditions, c)
+	}
+
+	if !failedTime.Inner.IsZero() {
+		c := apis.Condition{Type: apis.ConditionSucceeded, Status: v1.ConditionTrue, LastTransitionTime: failedTime}
+		pr.Status.Conditions = append(pr.Status.Conditions, c)
+	}
+
+	if ttl != nil {
+		pr.Spec.ExpirationSecondsTTL = ttl
+	}
+
+	return pr
+}
+
+func durationPointer(n int) *time.Duration {
+	s := time.Duration(n) * time.Second
+	return &s
+}
+
+func TestTimeLeft(t *testing.T) {
+	now := apis.VolatileTime{Inner: metav1.Now()}
+
+	PrTestCases := []struct {
+		name             string
+		completionTime   apis.VolatileTime
+		failedTime       apis.VolatileTime
+		ttl              *metav1.Duration
+		since            *time.Time
+		expectErr        bool
+		expectErrStr     string
+		expectedTimeLeft *time.Duration
+	}{
+		{
+			name:         "Error case: PipelineRun unfinished",
+			ttl:          &metav1.Duration{Duration: 10 * time.Second},
+			since:        &now.Inner.Time,
+			expectErr:    true,
+			expectErrStr: "should not be cleaned up",
+		},
+		{
+			name:           "Error case: PipelineRun completed now, no TTL",
+			completionTime: now,
+			since:          &now.Inner.Time,
+			expectErr:      true,
+			expectErrStr:   "should not be cleaned up",
+		},
+		{
+			name:             "PipelineRun completed now, 0s TTL",
+			completionTime:   now,
+			ttl:              &metav1.Duration{Duration: 0 * time.Second},
+			since:            &now.Inner.Time,
+			expectedTimeLeft: durationPointer(0),
+		},
+		{
+			name:             "PipelineRun completed now, 10s TTL",
+			completionTime:   now,
+			ttl:              &metav1.Duration{Duration: 10 * time.Second},
+			since:            &now.Inner.Time,
+			expectedTimeLeft: durationPointer(10),
+		},
+		{
+			name:             "PipelineRun completed 10s ago, 15s TTL",
+			completionTime:   apis.VolatileTime{Inner: metav1.NewTime(now.Inner.Add(-10 * time.Second))},
+			ttl:              &metav1.Duration{Duration: 15 * time.Second},
+			since:            &now.Inner.Time,
+			expectedTimeLeft: durationPointer(5),
+		},
+	}
+	for _, tc := range PrTestCases {
+		pr := newPipelineRun(tc.completionTime, tc.failedTime, tc.ttl)
+		d := test.Data{
+			PipelineRuns: []*apispipeline.PipelineRun{pr},
+		}
+		testAssets, cancel := getPipelineRunController(t, d)
+		defer cancel()
+		p, ok := testAssets.Controller.Reconciler.(*Reconciler)
```

Comment:
It'd be great to see the individual functions exercised in unit tests as well (which is hard when they're on the reconciler object - another +1 for moving the functions out of the reconciler

Reply:
The test needs to call prTimeLeft which is under reconciler object. I do this like reconciler/pipelinerun/pipelinerun_test and other unit tests. So...